### PR TITLE
feat: add backup detail and refactor backup

### PR DESCRIPTION
**Pull request by arinda-arif**
- change bigquery backup result table name to use timestamp suffix
- add extra information in backup list
- add backup detail sub command

### CODE REVIEW

**Comment by kushsharma** [comment]
https://github.com/odpf/proton/pull/66

**Comment by arinda-arif** [comment]
Hi @kushsharma, this PR is ready to be reviewed. Need your help on this 🙂

File: job/replay_manager.go [hunk 1]
```diff
@@ -126,6 +128,18 @@ func (m *Manager) Replay(ctx context.Context, reqInput models.ReplayRequest) (mo
...
 
+func prepareReplayConfig(allowedDownstreamNamespaces []string) map[string]string {
+	config := make(map[string]string)
+
+	ignoreDownstream := true
+	if len(allowedDownstreamNamespaces) > 0 {
+		ignoreDownstream = false
+	}
+	config[models.ConfigIgnoreDownstream] = strconv.FormatBool(ignoreDownstream)
+
```
Comment (kushsharma) [inline job/replay_manager.go:139]: We can rewrite this as
```go
config[models.ConfigIgnoreDownstream] = "true"
if len(allowedDownstreamNamespaces) > 0 {
	config[models.ConfigIgnoreDownstream] = "false"
}
```

File: mock/job.go [hunk 2]
```diff
@@ -52,6 +52,14 @@ func (repo *ProjectJobSpecRepository) GetAll(ctx context.Context) ([]models.JobS
 	return []models.JobSpec{}, args.Error(1)
 }
 
+func (repo *ProjectJobSpecRepository) GetAllWithNamespace(ctx context.Context) (map[string][]string, error) {
```
Comment (kushsharma) [inline mock/job.go:55]: Why are we adding this?

File: models/backup.go [hunk 3]
```diff
@@ -57,3 +57,11 @@ type BackupResult struct {
...
 	IgnoredResources []string
 }
+
+var (
+	// generic configurations
+	ConfigTTL              = "ttl"
+	ConfigIgnoreDownstream = "ignore_downstream"
+
+	BackupSpecKeyURN = "URN"
+)
```
Comment (kushsharma) [inline models/backup.go:67]: this could be `const` and defined at the top of the file?

File: models/datastore.go [hunk 4]
```diff
@@ -187,4 +187,5 @@ type DatastoreService interface {
 	BackupResourceDryRun(ctx context.Context, backupRequest BackupRequest, jobSpecs []JobSpec) (BackupPlan, error)
 	BackupResource(ctx context.Context, backupRequest BackupRequest, jobSpecs []JobSpec) (BackupResult, error)
 	ListBackupResources(ctx context.Context, projectSpec ProjectSpec, datastoreName string) ([]BackupSpec, error)
+	GetBackupResourceDetail(ctx context.Context, projectSpec ProjectSpec, datastoreName string, id uuid.UUID) (BackupSpec, error)
```
Comment (kushsharma) [inline models/datastore.go:190]: This should be named `GetResourceBackup` and similarly `ListBackupResources` as `ListResourceBackups`

**Review by kushsharma — CHANGES_REQUESTED** [review]

File: api/handler/v1/runtime.go [hunk 5]
```diff
@@ -1120,13 +1121,54 @@ func (sv *RuntimeServiceServer) ListBackups(ctx context.Context, req *pb.ListBac
...
+		return nil, status.Errorf(codes.Internal, "error while getting backup detail: %v", err)
+	}
+
+	var results []string
+	for _, result := range backupDetail.Result {
+		backupResult, ok := result.(map[string]interface{})
+		if !ok {
+			return nil, status.Errorf(codes.Internal, "error while parsing backup result: %v", ok)
+		}
+		results = append(results, backupResult[models.BackupSpecKeyURN].(string))
```
[the remark this answers is not in the document]
Comment (arinda-arif) [reply api/handler/v1/runtime.go]: I have added 2 error catch:
- if `models.BackupSpecKeyURN` doesn't exist in the map
- if the value cannot be casted to string
Please check if it is proper enough.

File: cmd/backup.go [hunk 6]
```diff
@@ -328,13 +330,111 @@ func printBackupListResponse(l log.Logger, listBackupsResponse *pb.ListBackupsRe
...
+			l.Info("unable to parse backup TTL")
+		}
+		expiry = backupDetailResponse.Spec.CreatedAt.AsTime().Add(ttlDuration)
+	}
+
+	table.Append([]string{"ID", backupDetailResponse.Spec.Id})
+	table.Append([]string{"Resource", backupDetailResponse.Spec.ResourceName})
+	table.Append([]string{"Created", backupDetailResponse.Spec.CreatedAt.AsTime().Format(time.RFC3339)})
+	table.Append([]string{"Ignore Downstream?", backupDetailResponse.Spec.Config[models.ConfigIgnoreDownstream]})
+	table.Append([]string{"Expired on", expiry.Format(time.RFC3339)})
```
[the remark this answers is not in the document]
Comment (arinda-arif) [reply cmd/backup.go]: Modified

File: mock/job.go (hunk 2 above)
Comment (arinda-arif) [reply mock/job.go]: Removed this. Missed to delete this one previously.

File: models/backup.go (hunk 3 above)
Comment (arinda-arif) [reply models/backup.go]: Updated

File: models/datastore.go (hunk 4 above)
Comment (arinda-arif) [reply models/datastore.go]: Updated

File: job/replay_manager.go (hunk 1 above)
Comment (arinda-arif) [reply job/replay_manager.go]: Updated

**Pull request merged**